### PR TITLE
Expose FixtureDef and SubRequest

### PR DESCRIPTION
**Pull request by asvetlov**
Export names required by `pytest_fixture_setup` and `pytest_fixture_post_finalizer` hook specs.

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
@bluetech im wondering, would there be a simple way to type the Sub-request with a FixtureParameter type as well

the idea bein that one coud use a type for the fixture param instead of any

**Comment by asvetlov** [comment]
@RonnyPfannschmidt `FixtureParameter` doesn't exist on the current `main` branch.
I assume I mean `FixtureValue`.

Now `SubRequest` is initialized with `fixturedef=FixtureDef[object]`.
Sure, we can derive `SubRequest` from `Generic[FixtureValue]` and change contructor's argument to `fixturedef=FixtureDef[FixtureValue]`.

If you mean something like this -- I can prepare a separate pull request.

The PR for exposing new public name is orthogonal to type system improvement and can be approved/merged independently.

File: changelog/9510.feature.rst [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Fixture types are now exported so the may be used in pytest plugin hooks.
```
Comment (bluetech) [inline changelog/9510.feature.rst:1]: Let's make the changelog refer to issue #7469 instead of the PR.

**Review by bluetech — CHANGES_REQUESTED** [review]
Hi @asvetlov,

Regarding `SubRequest`, please see https://github.com/pytest-dev/pytest/issues/8763#issuecomment-896130700. The situation hasn't changed since then unfortunately, it needs work.

Regarding `FixtureDef`, it needs a bit of an audit before exposing it. From looking at it now:

- The docstring "A container for a factory definition." is a bit obscure, we should improve it (but that's preexisting, not blocking).

- Need to decide if we want the constructor to be public or private. If it is public and exposed in `pytest`, that means pytest is committing to keeping it stable/backward compatible. Also, in this case the ctor parameters need to be documented.

  The conservative approach would be to make it private for now. Our mechanism for doing it is described [here](https://github.com/pytest-dev/pytest/blob/fe1f0e5376ce130d898b2f25fed167a9ce850867/src/_pytest/deprecated.py#L107). But before doing so we should check if there are many plugins which construct `FixtureDef`. (I have a large pytest plugins corpus which I use to check for such things, but I'm not on that machine ATM).

- Should document the non-underscored attributes, if they should be public, or make them private, if not. Need to audit them...

- Audit the methods:
   - `scope` -> Good
   - `addfinalizer` -> Good, should add a docstring
   - `finish` and `execute` -> Not sure about these. They also need `SubRequest` to be public. Need to see if/how the fixtures are meant to be using these.
   - `cache_key` -> I think it's probably better making it private. Might need a deprecation.

**Comment by bluetech** [comment]
> @RonnyPfannschmidt FixtureParameter doesn't exist on the current main branch. I assume you mean FixtureValue.

There are two things:

First, `FixtureValue`, also present on `FixtureRequest`, which is the type of the return value of the fixture.

@RonnyPfannschmidt is talking about `request.param`, which is the value of the parameter in a parametrized fixture. It is present only on `SubRequest` and is currently not typed. (Additional complication is that it's not always *set* - only for parametrized fixtures).

**Comment by RonnyPfannschmidt** [comment]
Hence the idea of the adding a generic for the case of having a parameter and a normal type when not

**Comment by bluetech** [comment]
`FixtureDef` is exposed now with making the ctor private and minimal exposing of internals (fb55615d5e999dd44306596f340036c195428ef1).

**Comment by bluetech** [comment]
I'll close this since just exposing SubRequest as-is is not going to be good.

**Pull request closed without merging**